Game/MyProject/Game/ModelLoader/ModelLoader.cpp

```diff
@@ -0,0 +1,129 @@
+}
+
+// --------------------------------
+// PRIVATE FUNCTION(S) AND METHOD(S)
+// --------------------------------
+
+// transforms the assimp scene node to my scene node
+void ModelLoader::processNode(aiNode* node, const aiScene* scene, SceneNode* my_scene_node, const std::string& shader_path) {
+    for (unsigned int i = 0; i < node->mNumMeshes; i++) {
+        aiMesh *mesh = scene->mMeshes[node->mMeshes[i]];
+        my_scene_node->addChild(makeSceneNode(mesh, scene, node, shader_path));
+    }
+    
+    for (unsigned int i = node->mNumMeshes; i < node->mNumChildren; i++)
+        my_scene_node->addChild(makeSceneNode(nullptr, scene, node, shader_path));
+    
+    for (unsigned int i = 0; i < node->mNumChildren; i++)
+        processNode(node->mChildren[i], scene, my_scene_node->getChild(i), shader_path);
+}
+
+// returns a new scene node
+std::shared_ptr<SceneNode> ModelLoader::makeSceneNode(const aiMesh *mesh, const aiScene *scene, const aiNode* node, const std::string& shader_path) {
+    std::shared_ptr<Mesh>           my_mesh;
+    std::shared_ptr<MeshInstance>   mesh_instace;
+    std::shared_ptr<SceneNode>      scene_node;
+    
+    std::vector<IvTCPVertex>        vertices;
+    std::vector<unsigned int>       indices;
+    
+    aiVector3t<float>               scale;
+    aiVector3t<float>               rotation;
+    aiVector3t<float>               position;
+    
+    scene_node = std::make_shared<SceneNode>(node->mName.data);
+    
+    if (mesh) {
+        for (unsigned int i = 0; i < mesh->mNumVertices; i++) {
+            IvTCPVertex vertex;
+            
+            if (mesh->HasPositions())
+                vertex.position = {mesh->mVertices[i].x, mesh->mVertices[i].y, mesh->mVertices[i].z};
+            
+            if (mesh->mTextureCoords[0])
+                vertex.texturecoord = {mesh->mTextureCoords[0][i].x, mesh->mTextureCoords[0][i].y};
+            
+            vertices.push_back(vertex);
+        }
+        
+        if (mesh->HasFaces()) {
+            for (unsigned int i = 0; i < mesh->mNumFaces; i++) {
+                
+                aiFace face = mesh->mFaces[i];
+                
+                for (unsigned int j = 0; j < face.mNumIndices; j++)
+                    indices.push_back(face.mIndices[j]);
+            }
+        }
+        
+        my_mesh = std::make_shared<Mesh>();
+        mesh_instace = std::make_shared<MeshInstance>();
+        
+        my_mesh->setVertexBuffer(vertices, kTCPFormat);
+        my_mesh->setIndexBuffer(indices);
+        mesh_instace->setMesh(my_mesh);
+        mesh_instace->setShader(shader_path.c_str());
+        scene_node->setRenderable(mesh_instace);
+        
+        if (mesh->mMaterialIndex >= 0) {
+            aiMaterial* material = scene->mMaterials[mesh->mMaterialIndex];
+    
+            aiString str;
+            material->GetTexture(aiTextureType_DIFFUSE, mesh->mMaterialIndex, &str);
+            if (str.length > 0) {
```

Comment:
Checking for null string is ok but there is no fallback option. In case the texture is missing or for whatever reason it can't be loaded the shader should have a "fallback" texture assigned (pink color , checkerboard, etc) . This is a good practice to see what textures are not loaded on the geometry and it also avoids the problem with unset uniforms (undefined behavior)